# Model Based Collaborative Filtering

<h3 style="color:#FA5F55;">Using Singular Value Decomposition (SVD) </h3>
<hr >

Memory-based recommendation systems are not always as fast and scalable as we would like them to be and are memory intensive, especially in the context of actual systems that generate real-time recommendations on the basis of very large datasets. To achieve these goals, model-based recommendation systems are used.

Model-based recommendation systems involve building a model based on the dataset of ratings. In other words, we extract some information from the dataset, and use that as a "model" to make recommendations without having to use the complete dataset every time. This approach offers the benefits of both speed and scalability.

Below is a simple implementation of Model based system using SVD.

In [3]:
!pip install surprise --q

In [13]:
# Importing necessary packages
from surprise import SVD
from surprise import Dataset
from surprise import accuracy
from surprise.model_selection import train_test_split

#### Surprise
In Surprise, there are three models: SVD, SVDpp, and NMF, out of which we employ SVD. NMF is a simplified version, ignoring user and item biases. SVDpp adds a feature where you also separately keep track of whether the user rated the item or not.

In [14]:
# Load the movielens dataset
data = Dataset.load_builtin('ml-100k')

In [15]:
# Splitting data for training and testing
trainset, testset = train_test_split(data, test_size=.20)

In [16]:
# Initializing SVD Algorithm
# Breaks into three matrices
model = SVD()

In [17]:
# Fitting the model
model.fit(trainset)

In [18]:
# Making predictions
predictions = model.test(testset)

# Computing error in terms of RMSE
accuracy.rmse(predictions)

RMSE: 0.9368


0.9368492335327254

In [19]:
# Predicting rating for a user for a movie
# uid = user id, iid = item/movie id
uid = str(456)
iid = str(131)

In [20]:
# Predicting for specific users and items
pred = model.predict(uid, iid)

In [21]:
# View predicting rating
pred.est

3.9418022469094134

This means that the user with uid=456 will rate movie with iid=131 as 3.9, which means we may recommend this movie to the user.

#### Other approaches
Within model based filtering, the model building process may be performed by different machine learning algorithms such as Bayesian network, clustering, and rule-based approaches. 

- The Bayesian network model formulates a probabilistic model for collaborative filtering problem. 

- The clustering model treats collaborative filtering as a classification problem and works by clustering similar users in same class and estimating the probability that a particular user is in a particular class C, and from there computes the conditional probability of ratings. 

- The rule-based approach applies association rule discovery algorithms to find association between co-purchased items and then generates item recommendation based on the strength of the association between items